vmdb/app/models/metric/ci_mixin/capture/openstack_base.rb

```diff
@@ -37,134 +37,364 @@ def perf_release_openstack
+        diff_metrics_avg = {}
+        i[:openstack_counters].each { |c| diff_metrics_avg[c] = metrics[c][:avg] - last_metrics[c][:avg] }
+
+        value = i[:calculation].call(diff_metrics_avg.values.first, intervals.values.first)
+      else
+        value = i[:calculation].call(metrics.values.first[:avg], nil)
+      end
+
+      # For (temporary) symmetry with VIM API we create 20-second intervals.
+      (last_period + 20.seconds..period).step_value(20.seconds).each do |ts_item|
+        counter_values_by_ts.store_path(ts_item.iso8601, i[:vim_style_counter_key], value)
+      end
+    end
+  end
 
-        # if the interval is not divisible by 20, keep an aggregate interval
-        # the two main risks here are:
-        #   1. the openstack admin configured a crazy interval like 17sec
-        #   2. one of the configured intervals contains an outlier that is
-        #      "sanded down" by the surrounding aggregated values
-        if interval % 20 != 0
-          aggregate << value
-          interval = ts - agg_start
-          # if the aggregate interval is divisible by 20,
-          #   average over the aggregate
-          #   reset the last_ts for the VIM API symmetry below
-          #   reset the agg_start
-          if interval % 20 == 0
-            value = aggregate.inject(0.0) { |sum, el| sum + el } / aggregate.size
-            last_ts = agg_start
-            agg_start = ts
+  def process_multi_counter_stats!(counter_values_by_ts, metric_capture_module, i, timestamps, metrics_by_counter_name,
+                                   data_collecting_period, log_header)
+    # !!! This method modifies counter_values_by_ts
+    # We have more counters in calculation. We have to make sure all counters have values present. It can
+    # happen that data of related counters are not collected in the same 20s window. So we will try to collect
+    # all of the data for each counter.
+    # Important Facts:
+    # 1. It can happen, that we will not have all samples for all counters, when we are near the borders of the
+    # collecting timeframe. We will attempt to find related data half of the data_collecting_period far. That
+    # should help us to avoid joining with the data from 2 Ceilometer collecting periods. If we will not find all
+    # data samples of all the counters, it's better to throw the data away, than storing not precise value.
+    # 2. If we will throw some incomplete data away, it's very likely it will be collected in next collecting
+    # period, because start_time is always moved one data_collecting_period back, to collect the pieces.
+    # 3. If collecting of the data takes longer, than half of the data_collecting_period, scale your Ceilometer.
+    # 4. Make sure the related counters are collected with same interval in Ceilometer pipeline, or this
+    # algorithm might not be able to match them together.
```

Comment:
I feel like there's tons of really good documentation in here.  Not only capturing why the code is written the way it is, but also comments on how to react to certain circumstances.

After this is all reviewed and merged, can you capture that in the guides section, or even in talk.manageiq.org?  We can go through what pieces are important to know outside the code.